Update move sec docs (#396)
aptos-labs/developer-docs: apps/docusaurus/docs/move/move-on-aptos/move-security-guidelines.md

---

### Object Ownership Check

Every `Object<T>` can be accessed by anyone, which means any `Object<T>` can be passed to any function, even if the caller doesn't own it.

> It's important to verify that the `signer` is the rightful owner of the object.

Reviewer: This isn't really any different than a resource as well?  Maybe we should mention that

#### Example Insecure Code

In this example, a user must make a payment before being able to perform actions within the module. The user needs to invoke `buy_action` to purchase an `Object<Action>`, which they can then use for executing operations later on.

```move
module user::obj{

    public entry fun buy_action(user:&signer,amount_of_action: u64){
        let price = calc_price(amount_of_action);
        deposit(user,price);

        let user_address = address_of(user);
        let constructor_ref = object::create_object(user_address);
        let object_signer = object::generate_signer(&constructor_ref);

>         move_to(&object_signer, Action {amount_of_action });

Reviewer 2: shouldn't this be `object::transfer()`?

    }

>     public entry fun perform_action(user:&signer,obj: Object<Action>){

Reviewer 2: it's not clear what this code does or why it matters that the ownership of the object is checked.

### Smart contract publishing key management

Using the same account for testnet and mainnet poses a security risk, as testnet private keys, often stored in less secure environments (ex. laptops), can be more easily exposed or leaked. An attacker that can obtain the private key for the testnet smart contract would be able to upgrade the mainnet one.

## Randomness

---

### Randomness - test-and-abort

> If a `public` function directly or indirectly invokes the randomness API, a malicious user can abuse the composability of this function and abort the transaction if the result is not as desired. This allows the user to keep trying until they achieve a beneficial outcome, undermining the randomness.

Reviewer 2: isn't this prevented at compile time?

Reviewer 3: Yes, https://github.com/aptos-labs/aptos-core/pull/13294

Reviewer 3: May we nonetheless document this as an invariant?

#### Example Vulnerable code

> ```move
> module user::lottery {
>     fun mint_to_user(user: &signer){
>         move_to(address_of(user),WIN{});
>     }
> 
>     #[randomness]
>     public entry fun play(user: &signer) {
>         let random_value = aptos_framework::randomness::u64_range(0, 100);
>         if (random_value == 42) {
>             mint_to_user(user);
>         }
>     }

Reviewer: Can we stay concrete and the same for these examples?

Here we use WIN, and below we use Object_for_the_winner
```suggestion
```move
module user::lottery {
    struct Winner {}

    fun mint_to_user(user: &signer) {
        move_to(address_of(user), Winner{});
    }

    #[randomness]
    public entry fun play(user: &signer) {
        let random_value = aptos_framework::randomness::u64_range(0, 100);
        if (random_value == 42) {
            mint_to_user(user);
        }
    }
```

```move
module attacker::exploit {
    public entry fun exploit(attacker: &signer) {
        @user::lottery::play(attacker);
        assert!(exists<@user::lottery::WIN>(address_of(attacker)));
    }
}
```

To resolve the possible issue, is sufficient to set the visibility of all functions that invoke the randomness API, either directly or indirectly, to `entry` rather than `public` or `public entry`.

#### Example Secure Code

```move
module user::lottery {
    fun mint_to_user(user: &signer){
        move_to(address_of(user),Object_for_the_winner{});
    }

    #[randomness]
    entry fun play(user: &signer) {
        let random_value = aptos_framework::randomness::u64_range(0, 100);
        if (random_value == 42) {
            mint_to_user(user);
        }
    }
}
```

> At Aptos, We are always security-first. During compilation, we ensure that no randomness API is invoked from a public function. However, we still allow users to make this choice by adding the attribute `#[lint::allow_unsafe_randomness]` to the public function.

### Randomness - undergasing

> When different code paths in a function consume different amounts of gas, an attacker can manipulate the gas limit to bias the outcome.
> 
> If the "lose" path consumes more gas than the "win" path, an attacker can set the gas limit such that the "lose" path will run out of gas. The transaction will be aborted and the user will never lose.

Reviewer: This is repeated in the paragraph below, let's clean this up
```suggestion
When different code paths in a function consume different amounts of gas, an attacker can manipulate the gas limit to bias the outcome.  Let's look at an example of how different paths can consume different amounts of gas.
```

#### Example Vulnerable code

```
module user::lottery {
    fun win(user: &signer) {
        [...]
    }

    fun lose(user: &signer) {
        [...]
    }
    #[randomness]
    entry fun play(user: &signer) {
        let random_value = aptos_framework::randomness::u64_range(0, 100);
        if (random_value == 42) {
            win(user);
        } else {
            lose(user);
        }
    }
}
```

> In this lottery-example, `win` and `lose` consume different amounts of gas.

Reviewer 2: from the example it's not clear why they consume different amounts of gas.

> If `lose` consumes more gas than `win`, an attacker can set the gas limit that is sufficient for `win` but not for `lose`. This forces the transaction to abort when the "lose" path is taken, ensuring that the user never submits a valid transaction for the "lose" path.

Reviewer: ```suggestion
If the `lose` function consumes more gas than the `win` function, an attacker can set the max gas limit that is sufficient for `win` but not for `lose`. This forces the transaction to abort when the `lose` path is taken, ensuring that the user will never execute the `lose` path.  Then, the user can call the function repeatedly until they win.
```

#### Example Secure Code

There are different ways to secure the code:

1. Make sure better outcomes use more gas.
2. Allows only admin addresses to invoke randomness API.
3. Make entry functions work regardless of random outcomes. Avoid immediate actions based on randomness for consistent gas use.

> > We will be providing more functionality in the future, to allow for more complex code to be able to be safe against undergasing attacks.

Reviewer 2: should we add a link to the complete randomness documentation?

Reviewer: +1 please add appropriate randomness docs link here

Author: Added at the beginning of the randomness section

```move
module attacker::exploit {
    entry fun exploit(attacker: &signer) {
        @user::lottery::play(attacker);
        assert!(exists<@user::lottery::WIN>(address_of(attacker)));
    }
}
```

To resolve the possible issue, is sufficient to set the visibility of all functions that invoke the randomness API, either directly or indirectly, to `entry` rather than `public` or `public entry`.

#### Example Secure Code

```move
module user::lottery {
    fun mint_to_user(user: &signer) {
        move_to(user, WIN {});
    }

    #[lint::allow_unsafe_randomness]
    entry fun play(user: &signer) {
        let random_value = aptos_framework::randomness::u64_range(0, 100);
        if (random_value == 42) {
            mint_to_user(user);
        }
    }
}
```

### Randomness - undergasing

When different code paths in a function consume different amounts of gas, an attacker can manipulate the gas limit to bias the outcome. Let's look at an example of how different paths can consume different amounts of gas.

#### Example Vulnerable code

> ```

Reviewer 4: ```suggestion
```move filename="randomness_example.move"
```
